src/maxtree.jl

```diff
@@ -0,0 +1,881 @@
+`min_diameter` with the brighter value taken from their first ancestral component
+(in *max-tree* representation of `image`) that has the diameter no smaller than
+`min_diameter`.
+
+# Arguments
+- `image::AbstractArray`: the ``N``-dimensional input image
+- `min_diameter::Number=8`: the minimal length (in pixels) of the widest
+  dimension of the bounding box of the image component to keep intact
+- `connectivity::Integer=1`: the neighborhood connectivity. The maximum number
+  of orthogonal steps to reach a neighbor of the pixel. In 2D, it is 1 for
+  a 4-neighborhood and 2 for a 8-neighborhood.
+- `maxtree::MaxTree=nothing`: optional pre-built *max-tree*. Note that,
+  when `maxtree` is specified, the `connectivity` option is ignored.
+
+# Returns
+An array of the same type and shape as the `image`.
+
+# See also
+[`diameter_closing!`](@ref), [`diameter_opening`](@ref),
+[`area_closing`](@ref),
+[`MaxTree`](@ref), [`closing`](@ref)
+
+# References
+1. Walter, T., & Klein, J.-C. (2002). *Automatic Detection of Microaneurysms in Color Fundus Images of the Human Retina by Means of the Bounding Box Closing*. In A. Colosimo, P. Sirabella, A. Giuliani (Eds.), *Medical Data Analysis. Lecture Notes in Computer Science*, vol 2526, 210-220. Springer Berlin Heidelberg.
+   > https://doi.org/10.1007/3-540-36104-9_23
+2. Carlinet, E., & Geraud, T. (2014). *A Comparative Review of Component Tree Computation Algorithms*. IEEE Transactions on Image Processing, 23(9), 3885-3895.
+   > https://doi.org/10.1109/TIP.2014.2336551
+
+# Examples
+Creating a test image `f` (quadratic function with a minimum in the center and
+4 additional local minima):
+```jldoctest
+julia> w = 12;
+julia> f = [180 + 0.2*((x - w/2)^2 + (y-w/2)^2) for x in 0:w, y in 0:w];
+julia> f[3:4, 2:6] .= 40; f[3:5, 10:12] .= 60; f[10:12, 3:5] .= 80;
+julia> f[10:11, 10:12] .= 100; f[11, 11] = 100;
+```
+Area closing of `f`:
+```jldoctest
+julia> f_dclose = diameter_closing(f, min_diameter=3, connectivity=1)
+```
+All small minima with a diameter of 2 or less are removed.
+For the remaining minima the widest bounding box side is at least 3.
+"""
+diameter_closing(image::AbstractArray; kwargs...) =
+    diameter_closing!(similar(image), image; kwargs...)
+
+# Calculates the local maxima or minima of the image using the max-tree
+# representation (maxima if maxtree.rev=false, minima otherwise).
+# It's not the most effecient method for extrema calculation, but could be
+# useful if the max-tree is already calculated.
+# Each minima/maxima region is labeled with the unique id (1 is the global
+# maximum/minimum).
+function local_extrema!(output::AbstractArray{<:Any, N},
+                        image::AbstractArray{<:Any, N},
+                        maxtree::MaxTree{N}) where N
+    (axes(image) == axes(maxtree)) || throw(DimensionMismatch())
+    (size(output) == size(image)) || throw(DimensionMismatch())
+
+    next_label = one(eltype(output))
```

Comment:
```suggestion
    next_label = oneunit(eltype(output))
```

Reply:
Julia 1.0 complains here:
```
MethodError: no method matching CartesianIndex{2}(::CartesianIndex{2})
  Closest candidates are:
    CartesianIndex{2}() where N at multidimensional.jl:73
    CartesianIndex{2}(!Matched::Integer...) where N at multidimensional.jl:69
    CartesianIndex{2}(!Matched::Tuple{Vararg{Integer,N}}) where N at multidimensional.jl:64
    ...
  Stacktrace:
   [1] oneunit(::Type{CartesianIndex{2}}) at ./number.jl:321
   [2] neighbor_cartesian_offsets(::Type{CartesianIndex{2}}, ::Int64) at /home/travis/build/JuliaImages/ImageMorphology.jl/src/maxtree.jl:137
```
What would be the workaround?

Reply:
Copy the related codes from Julia souce codes in a new file (say `compat.jl`):

```julia
@doc (@doc oneunit)
_oneunit(x) = oneunit(x)
if v"1.0" <= VERSION < v"1.1"
    # copy from julia source codes for 1.0 compatibility
    _oneunit(::CartesianIndex{N}) where {N} = oneunit(CartesianIndex{N})
    _oneunit(::Type{CartesianIndex{N}}) where {N} = CartesianIndex(ntuple(x -> 1, Val(N)))
end
```

and replace incompatible `oneunit` by `_oneunit`

Reply:
In Julia 2.0 `one(CartesianIndex{N})` will throw an error, so maybe just put a comment on that line.